torchdata/datapipes/iter/util/cacheholder.py

```diff
@@ -68,70 +68,107 @@ def __len__(self) -> int:
+
+        cache_dp = url.on_disk_cache(extra_check_fn=hash_fn).open_url().map(fn=lambda x: b''.join(x), input_col=1).end_caching()
     """
-
-    def __new__(
-        cls,
-        source_datapipe: IterDataPipe,
-        opDataPipe,
-        op_args=None,
-        op_kwargs=None,
-        op_map: Optional[Callable] = None,
-        mode: str = "wb",
-        filepath_fn: Callable = _default_filepath_fn,
+    def __init__(
+        self,
+        source_datapipe,
+        filepath_fn=_default_filepath_fn,
+        extra_check_fn=None,
     ):
-
-        assert isinstance(source_datapipe, IterDataPipe), "'source_datapipe' needs to be an IterDataPipe"
-        if op_args is None:
-            op_args = ()
-        if op_kwargs is None:
-            op_kwargs = {}
-        # source_datapipe should either yield url or file string
-        dp = source_datapipe.map(fn=lambda data: (filepath_fn(data), data))
-        # Out of order
-        cached_dp, todo_dp = dp.demux(2, lambda d: 0 if path.exists(d[0]) else 1)
+        self.source_datapipe = source_datapipe
+        self.filepath_fn = filepath_fn
+        self.extra_check_fn = extra_check_fn
+        self._end_caching_flag: bool = False
+
+    def __iter__(self):
+        if self._end_caching_flag:
+            for d in self.source_datapipe:
+                yield d
+        else:
+            # In case of BC breaking, use RuntimeError for now.
+            # Can be a warning
+            raise RuntimeError("Please call `end_caching()` before iteration.")
+
+    def __add__(self, other_datapipe):
+        raise RuntimeError("`OnDiskCacheHolder` doesn't support add operation")
+
+    def _cache_check_fn(self, data):
+        filepath = self.filepath_fn(data)
+        if not os.path.exists(filepath):
+            return False
+        if self.extra_check_fn:
+            return self.extra_check_fn(filepath)
+        return True
+
+    def _end_caching(self):
+        todo_dp, cached_dp = self.source_datapipe.demux(2, self._cache_check_fn)
+        self._end_caching_flag = True
+        # Re-assign source_datapipe
+        self.source_datapipe = todo_dp
 
         # Cached: keeps filepath
-        cached_dp = cached_dp.map(fn=lambda data: data[0])
-
-        # Non-cached
-        # opDataPipe yield bytes
-        if not op_map:
-            todo_dp = opDataPipe(todo_dp.map(fn=lambda data: data[1]), *op_args, **op_kwargs).save_to_disk(
-                mode=mode, filepath_fn=filepath_fn
-            )
-        else:
-            todo_dp = (
-                opDataPipe(todo_dp.map(fn=lambda data: data[1]), *op_args, **op_kwargs)
-                .map(op_map)
-                .save_to_disk(mode=mode, filepath_fn=filepath_fn)
-            )
-
-        # Load files from local disk
-        return FileLoader(cached_dp.concat(todo_dp))
+        return cached_dp.map(fn=self.filepath_fn)
+
+
+@functional_datapipe("end_caching")
+class EndOnDiskCacheHolderIterDataPipe(IterDataPipe):
```

Comment:
It's too hard for me to find a proper name of this. Please feel free to provide any suggestion.